### PR TITLE
Improvements to MultiCursor

### PR DESCRIPTION
**Pull request by mementum**
MultiCursor as currently implemented is always "multi" in that it shows in all axes and there are no options to control that.

Additionally it will distort axes, because the initial calculation of the position of the lines is based on the last passed axes and not on each individual axes.

The pull request adds additional kwargs to control the behavior
- horizMulti (default -> False): if the horizontal line has to be show on all axes at the same time
- vertMulti (default -> True): if the vertical line has to be shown on all axes at the same time

The combination of the above creates a "crosshair" effect in which the horizontal line shows up in the axes only when the mouse is over it, but the vertical line is always shown. Given that most plots are created with a common x axis and different y axis, the default behavior seems logical.
- horizShared (default -> True): if the horizontal axis is shared the position of the vertical lines will be calculated from a single axis (the last one), else an individual calculation per axis will happen
- vertShared (default -> False): if the vertical axis is shared the position of the horizontal lines will be calculated from a single axis (the last one), else an individual calculation per axis will happen

The defaults follow also the above logic about the most common type of charts.

The patch also clears some apparent "whitespace" leftovers.

### CODE REVIEW

File: lib/matplotlib/widgets.py [hunk 1]
```diff
@@ -1110,24 +1124,36 @@ def onmove(self, event):
             return
         if self.vertOn:
             for line in self.vlines:
-                line.set_xdata((event.xdata, event.xdata))
-                line.set_visible(self.visible)
+                visible = self.visible
+                if not self.vertMulti:
+                    visible = visible and line.axes == event.inaxes
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: `is` might be a better choice of test here?

File: lib/matplotlib/widgets.py [hunk 2]
```diff
@@ -1110,24 +1124,36 @@ def onmove(self, event):
...
 
-    def _update(self):
+    def _update(self, event):
         if self.useblit:
             if self.background is not None:
                 self.canvas.restore_region(self.background)
             if self.vertOn:
                 for ax, line in zip(self.axes, self.vlines):
-                    ax.draw_artist(line)
+                    if self.vertMulti or event.inaxes == line.axes:
```
Comment (tacaswell) [inline lib/matplotlib/widgets.py]: There is only one background for the whole canvas, we need to redraw all visible lines, not just the ones in the current axes.

**Comment by tacaswell** [comment]
Sorry this sat un-reviewed for so long :disappointed: .

Modulo a few small comments this seems reasonable.  Can you change the defaults so the the current behavior is un-changed?

This will need to be documented both in the docstring and with a `whats_new` entry. 

We might also want to bikeshed those new kwarg names a bite.

File: lib/matplotlib/widgets.py (hunk 1 above)
Comment (efiring) [inline lib/matplotlib/widgets.py]: And I don't understand the point of the `visible = visible` part.

Comment (efiring) [inline lib/matplotlib/widgets.py]: OK, now I see--that red "and" in the middle fooled my eye into splitting this in the wrong place.

**Comment by tacaswell** [comment]
@mementum Closing this as it has sat for over a year with no action.

Please ping if you are still interested in working on this!

**Pull request closed without merging**